File: vuln-sandbox/go/cwe-918/ssrf.go
// VULNERABLE: go-http-get-sprintf — http.Get with fmt.Sprintf URL from user input
// Rule: go-http-get-sprintf | CWE-918 | Severity: HIGH

package main

import (
	"fmt"
	"io"
	"net/http"
)

func proxyHandler(w http.ResponseWriter, r *http.Request) {
	userHost := r.URL.Query().Get("host")

	// VULNERABLE: user-controlled host used to construct outbound HTTP request
	// An attacker can pass: 169.254.169.254/latest/meta-data to access cloud metadata
	resp, err := http.Get(fmt.Sprintf("http://%s/api/data", userHost))
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadGateway)
		return
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	fmt.Fprintf(w, "%s", body)
}

func main() {
	http.HandleFunc("/proxy", proxyHandler)
	http.ListenAndServe(":8080", nil)
}

File: vuln-sandbox/typescript/cwe-918/ssrf.ts
// VULNERABLE: SsrfFetchUserInput — fetch with user-controlled URL (TypeScript)
// Rule: SsrfFetchUserInput | CWE-918 | Severity: HIGH

import express, { Request, Response } from 'express';

const app = express();

app.get('/proxy', async (req: Request, res: Response): Promise<void> => {
  const targetUrl: string = req.query.url as string;

  try {
    // VULNERABLE: user-controlled URL passed directly to fetch
    // An attacker can pass: http://169.254.169.254/latest/meta-data to access cloud metadata
    const response = await fetch(targetUrl);
    const data = await response.text();
    res.send(data);
  } catch (err) {
    res.status(502).json({ error: 'Bad gateway' });
  }
});

app.listen(3000);
